autoprotocol/protocol.py

```diff
@@ -5080,51 +5081,62 @@ def provision(self, resource_id, dests, volumes):
+            if not isinstance(amount, (str, Unit)):
+                raise TypeError("Volume/Mass must be a string or Unit.")
+            if amount.unit in VOLUMETRIC_UNITS:
+                unique_measure_modes.add("volume")
+            elif amount.unit in MASS_UNITS:
+                unique_measure_modes.add("mass")
+
+        if len(unique_measure_modes) != 1:
+            raise ValueError("Received amounts with more than one measurement mode")
+        measurement_mode = unique_measure_modes.pop()
         provision_instructions_to_return: Provision = []
-        for d, v in zip(dests, volumes):
-            d_max_vol = d.container.container_type.true_max_vol_ul
-            if v > d_max_vol:
-                raise ValueError(
-                    f"The volume you are trying to provision ({v}) exceeds the "
-                    f"maximum capacity of this well ({d_max_vol})."
-                )
-            self._remove_cover(d.container, "provision")
-            dest_group = []
-            if v > Unit(900, "microliter"):
-                diff = v - Unit(900, "microliter")
-                provision_instructions_to_return.append(
-                    self.provision(resource_id, d, Unit(900, "microliter"))
-                )
-                while diff > Unit(0.0, "microliter"):
-                    provision_instructions_to_return.append(
-                        self.provision(resource_id, d, diff)
-                    )
-                    diff -= diff
-                continue
+
+        for d, amount in zip(dests, amounts):
 
             xfer = {}
             xfer["well"] = d
-            xfer["volume"] = v
+            xfer[measurement_mode] = amount
 
-            if d.volume:
-                d.volume += v
-            else:
-                d.set_volume(v)
-            dest_group.append(xfer)
+            if measurement_mode == "volume":
+                d_max_vol = d.container.container_type.true_max_vol_ul
+                if amount > d_max_vol:
+                    raise ValueError(
+                        f"The volume you are trying to provision ({amount}) exceeds the "
+                        f"maximum capacity of this well ({d_max_vol})."
+                    )
+                if amount > Unit(900, "microliter"):
+                    diff = amount - Unit(900, "microliter")
+                    provision_instructions_to_return.append(
+                        self.provision(resource_id, d, Unit(900, "microliter"))
+                    )
+                    while diff > Unit(0.0, "microliter"):
+                        provision_instructions_to_return.append(
+                            self.provision(resource_id, d, diff)
+                        )
+                        diff -= diff
+                    continue
+
+                if d.volume:
+                    d.volume += amount
+                else:
+                    d.set_volume(amount)
+            self._remove_cover(d.container, "provision")
```

Comment:
remove cover call should happen before we start provisioning

Reply:
Good catch.